docs/design/Refactor Azure authentication.md

```diff
@@ -0,0 +1,165 @@
+# **Azure Authentication Refactoring in Ratify**
+
+## **Introduction**
+Authentication is a critical process in Ratify, ensuring secure access to artifatcs in container registries, and to keys, secrets and certificates from cloud key vaults, and other resources. Azure offers two primary SDKs for authentication in Go:
+
+- **Azure Identity ([azidentity](https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication?tabs=bash))**: Designed for seamless integration with Azure services.
+- **Microsoft Authentication Library ([MSAL](https://learn.microsoft.com/en-us/entra/identity-platform/msal-overview))**: Provides advanced token management capabilities.
+
+Currently, Ratify uses both SDKs across different components, leading to complexity and maintenance overhead. This document proposes a comprehensive refactoring of Azure authentication in Ratify to improve maintainability, reduce duplication, and streamline the user experience.
+
+---
+
+## **Existing Azure Authentication in Ratify**
+
+### **ACR Token Retrieval**
+Located in the **ORAS auth providers (`pkg/common/oras/authprovider/azure`)**:
+1. **Azure Managed Identity (`azureidentity.go`)**:
+   - Uses `azidentity.NewManagedIdentityCredential` to retrieve an access token.
+   - Requires only the `clientID`:
+     ```go
+     id := azidentity.ClientID(clientID)
+     opts := azidentity.ManagedIdentityCredentialOptions{ID: id}
+     cred, err := azidentity.NewManagedIdentityCredential(&opts)
+     ```
+
+2. **Azure Workload Identity (`azureworkloadidentity.go`)**:
+   - Uses `confidential.NewCredFromAssertionCallback` from the **MSAL** package. 
+
+### **Key Management Provider and Certificate Provider**
+Both components recently replaced the deprecated `autorest` SDK with `azidentity` and now use workload identity credentials for authentication.
+
+---
+
+## **Challenges with the Current Design**
+
+### 1. **Multiple SDKs**
+Ratify employs both **`MSAL`** and **`azidentity`**, increasing the maintenance burden. Consolidating to a single SDK simplifies dependency management, reduces upgrade complexity, and enhances maintainability.
+
+### 2. **Code Duplication**
+Significant code duplication exists across components, particularly between Azure workload identity and managed identity implementations. Consolidating shared logic improves maintainability.
+
+### 3. **Explicit Authentication Selection**
+Currently, users must explicitly specify the authentication type. In well-defined environments like Azure Kubernetes Service (AKS), this should be inferred automatically based on environment variables.
+
+---
+
+## **Proposed Refactoring**
+
+### **Goals**
+1. Design a **common package** for Azure authentication logic.
+2. **Infer authentication type** automatically based on the environment, reducing user configuration overhead.
+3. **Unify implementations** for workload identity and managed identity in ORAS auth providers.
+4. Implement a **chained authentication process**:
```

Comment:
Could you elaborate why the `DefaultAzureCredential` is not sufficient?

Reply:
DefaultAzureCredential is similar to chained credential with one difference: it uses a specific chain of credentials. With Chained credential, we have the liberty of defining the priority of the credentials. In our case, I believe most of the use cases are going to be workload identity in an Azure Kubernetes cluster so that should be on top of the hierarchy. We then fall back to managed identity if workload identity fails, and similarly, fall back to CLI when managed identity fails.